### PR TITLE
Try to lookup in the shared dict when the lock times out

### PR DESCRIPTION
**Pull request by jdesgats**
This attempts to recover from fairly rare race condition that might cause some threads to exit in a timeout state while the value is actually available. I've seen it happening only under very high load and it was still pretty rare.

------------

Due to race conditions with other blocked requests in other worker
processes, the lock might time out because other processes took the
shared dict causing the `resty.lock` spin lock to wait again and again.

This just works around the problem by querying the shared dict for lock
times out. The real problem is that a dumb lock is not the right
synchronization tool here: a condition-style lock that releases all
stalled requests without requiring further locking would be more
efficient as it would prevent the unnecessary timeout at the first
place.

### CODE REVIEW

File: lib/resty/mlcache.lua [hunk 1]
```diff
@@ -670,7 +670,31 @@ function _M:get(key, opts, cb, ...)
...
+        if data == CACHE_MISS_SENTINEL_LRU then
+            return nil, nil, 2
+        end
+
+        if data ~= nil then
+            return data, nil, 2
+        end
+
+        -- ok, this time there's no more hope
+        return nil, "could not aquire callback lock: timeout"
```
Comment (thibaultcha) [inline lib/resty/mlcache.lua]: @jdesgats This is a great catch, thank you!

It seems to me like the following patch would achieve the same behaviour in a less invasive way: 
```diff
diff --git a/lib/resty/mlcache.lua b/lib/resty/mlcache.lua
index 35b0db18..ec3ae039 100644
--- a/lib/resty/mlcache.lua
+++ b/lib/resty/mlcache.lua
@@ -607,7 +607,7 @@ end
 
 local function unlock_and_ret(lock, res, err, hit_lvl)
     local ok, lerr = lock:unlock()
-    if not ok then
+    if not ok and lerr ~= "unlocked" then
         return nil, "could not unlock callback: " .. lerr
     end
 
@@ -668,9 +668,9 @@ function _M:get(key, opts, cb, ...)
         return nil, "could not create lock: " .. err
     end
 
-    local elapsed, err = lock:lock(LOCK_KEY_PREFIX .. namespaced_key)
-    if not elapsed then
-        return nil, "could not aquire callback lock: " .. err
+    local elapsed, lerr = lock:lock(LOCK_KEY_PREFIX .. namespaced_key)
+    if not elapsed and lerr ~= "timeout" then
+        return nil, "could not aquire callback lock: " .. lerr
     end
 
     -- check for another worker's success at running the callback
@@ -680,14 +680,22 @@ function _M:get(key, opts, cb, ...)
         return unlock_and_ret(lock, nil, err)
     end
 
-    if data then
-        if data == CACHE_MISS_SENTINEL_LRU then
-            return unlock_and_ret(lock, nil, nil, 2)
-        end
+    if data == CACHE_MISS_SENTINEL_LRU then
+        return unlock_and_ret(lock, nil, nil, 2)
+    end
 
+    if data ~= nil then
         return unlock_and_ret(lock, data, nil, 2)
     end
 
+    -- data is nil, we are either the 1st worker to hold the lock, or
+    -- a subsequent worker whose lock has timed out before the 1st one
+    -- finished to run the callback.
+
+    if lerr == "timeout" then
+        return nil, "could not acquire callback lock: timeout"
+    end
+
     -- still not in shm, we are responsible for running the callback
 
     local pok, perr, err, new_ttl = pcall(cb, ...)
```

What do you think?

**Comment by thibaultcha** [comment]
> The real problem is that a dumb lock is not the right
> synchronization tool here: a condition-style lock that releases all
> stalled requests without requiring further locking would be more
> efficient as it would prevent the unnecessary timeout at the first
> place.

Indeed. We can probably write our own little helper for this purpose. It seems to me like a "dumber" lock (one that does not subsequently lock after the first unlock) would be fine here.

File: lib/resty/mlcache.lua (hunk 1 above)
Comment (jdesgats) [reply lib/resty/mlcache.lua]: Yeah it looks that you approach works as well in a cleaner way without adding even more branches to the already long `:get()` method :+1: 

Feel free to decline my PR and commit your fix.

Comment (thibaultcha) [reply lib/resty/mlcache.lua]: It is fine with me if you want to replace your patch with this one as well. If not, I will commit this patch tomorrow and credit you in the commit :+1:

**Comment by thibaultcha** [comment]
Patch committed to master. Thanks for the report!

**Pull request closed without merging**